### PR TITLE
IMPRO-1503 Pt 2 - Factor out forecast/truth splitting code

### PR DESCRIPTION
**Pull request by bayliffe**
Factors code out of the estimate emos coefficients CLI that is used to split forecast and truth cubes so this can be reused in the reliability calibration work.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- improver_tests/calibration/test_init.py  2
         

Complexity decreasing per file
==============================
+ improver_tests/acceptance/test_estimate_emos_coefficients.py  -1
+ improver/cli/estimate_emos_coefficients.py  -9
         

Clones added
============
- improver_tests/calibration/test_init.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4857190&bid=16114369)

File: improver_tests/calibration/test_init.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Unit tests for calibration.__init__"""
+
+import unittest
+import iris
+import numpy as np
+from datetime import datetime
```
Comment (benfitzpatrick) [inline improver_tests/calibration/test_init.py]: Codacy Issue found: [standard import "from datetime import datetime" should be placed before "import iris"](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4857190)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver_tests/calibration/test_init.py  2
         

Complexity decreasing per file
==============================
+ improver_tests/acceptance/test_estimate_emos_coefficients.py  -1
+ improver/cli/estimate_emos_coefficients.py  -9
         

Clones added
============
- improver_tests/calibration/test_init.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4857190&bid=16114369)

File: improver/calibration/__init__.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+# POSSIBILITY OF SUCH DAMAGE.
+"""init for calibration"""
+
+from collections import OrderedDict
+
+from improver.utilities.cube_manipulation import MergeCubes
+from improver.metadata.probabilistic import extract_diagnostic_name
+
+
+def split_forecasts_and_truth(cubes, truth_attribute):
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/calibration/__init__.py]: Added.

File: improver/calibration/__init__.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        (tuple): tuple containing:
+            **forecast** (iris.cube.Cube):
+                A cube containing all the historic forecasts.
+            **truth** (iris.cube.Cube):
+                A cube containing all the truth data.
+            **land_sea_mask** (iris.cube.Cube or None):
+                If found within the input cubes list a land-sea mask will be
+                returned, else None is returned.
+    Raises:
+        RuntimeError:
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/calibration/__init__.py]: Done.

File: improver/calibration/__init__.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+            An unexpected number of distinct cube names were passed in.
+        RuntimeError:
+            More than one cube was identified as a land-sea mask.
+        RuntimeError:
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/calibration/__init__.py]: Done.

File: improver_tests/calibration/test_init.py [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+"""Unit tests for calibration.__init__"""
+
+import unittest
+from datetime import datetime
+import numpy as np
+import iris
+
+from improver.calibration import split_forecasts_and_truth
+from improver_tests.set_up_test_cubes import (set_up_variable_cube,
+                                              set_up_probability_cube)
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver_tests/calibration/test_init.py]: Reordered.

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver_tests/calibration/test_init.py  2
         

Complexity decreasing per file
==============================
+ improver_tests/acceptance/test_estimate_emos_coefficients.py  -1
+ improver/cli/estimate_emos_coefficients.py  -9
         

Clones added
============
- improver_tests/calibration/test_init.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4857190&bid=16114369)

File: improver/calibration/__init__.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+def split_forecasts_and_truth(cubes, truth_attribute):
+    """
+    A common utility for splitting the various inputs cubes required for
+    calibration CLIs. These are generally the forecast cubes, historic truths,
+    and in some instances a land-sea mask is also required.
+
+    Args:
+        cubes (list):
+            A list of input cubes which will be split into relevant groups.
+            These include the historical forecasts, either as realization or
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/calibration/__init__.py]: Done.

File: improver_tests/calibration/test_init.py [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+
+    """Test the split_forecasts_and_truth method."""
+
+    def setUp(self):
+        """Create forecast and truth cubes for use in testing the reliability
+        calibration plugin. Two forecast and two truth cubes are created, each
+        pair containing the same data but given different forecast reference
+        times and validity times. These times maintain the same forecast period
+        for each forecast cube.
+        Each forecast cube in conjunction with the contemporaneous truth cube
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver_tests/calibration/test_init.py]: Nope, copy/paste error. Modified.

File: improver_tests/calibration/test_init.py [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+                         forecast.name())
+        self.assertEqual('air_temperature', truth.name())
+        self.assertEqual('land_binary_mask', land_sea_mask.name())
+        self.assertSequenceEqual((2, 2, 4, 4), forecast.shape)
+        self.assertSequenceEqual((2, 4, 4), truth.shape)
+        self.assertSequenceEqual((4, 4), land_sea_mask.shape)
+
+    def test_realization_data(self):
+        """Test that when multiple forecast cubes and truth cubes are provided,
+        he groups are created as expected."""
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver_tests/calibration/test_init.py]: Done.

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver_tests/calibration/test_init.py  2
         

Complexity decreasing per file
==============================
+ improver_tests/acceptance/test_estimate_emos_coefficients.py  -1
+ improver/cli/estimate_emos_coefficients.py  -9
         

Clones added
============
- improver_tests/calibration/test_init.py  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4857190&bid=16114369)

File: improver/calibration/__init__.py [hunk 9]
```diff
@@ -45,8 +45,8 @@ def split_forecasts_and_truth(cubes, truth_attribute):
     Args:
         cubes (list):
             A list of input cubes which will be split into relevant groups.
-            These include the historical forecasts, either as realization or
-            probability cubes, and the truth cubes.
+            These include the historical forecasts, in the format supported by
+            the calibration CLIs, and the truth cubes,
```
Comment (gavinevans) [inline improver/calibration/__init__.py]: Typo:
```suggestion
            the calibration CLIs, and the truth cubes.
```

**Pull request merged**